Fix comparison of a WriteConcern object with None.

pcorpet opened the pull request:
This would make errors like in this bug mongomock/mongomock#458 more readable.

prashantmital commented:
Thanks for the pull @pcorpet! A couple of suggestions:

- Instead of checking the truth value of `other`, we should ideally check its type and raise an exception if we encounter an unexpected type. This pattern is implemented in `ReadConcern.__eq__`: https://github.com/mongodb/mongo-python-driver/blob/dea14beb865731f64a43d6f66f96784d259307db/pymongo/read_concern.py#L65.

- It'd be great if you could add tests for `WriteConcern`, even if they just test the `__eq__` functionality. This could go into a new file such as `tests/test_write_concern.py`.

pcorpet commented:
Thanks @prashantmital for looking at my PR so quickly. I've added some tests, however for your other suggestion, I prefer to keep the comparison with objects that are compatible with WriteConcern but are not the exact same class. I am the maintainer of the https://github.com/mongomock/mongomock project which implements another WriteConcern compatible class and I would prefer to keep the comparison possible if that's OK with you. I've added corresponding tests.

prashantmital commented:
@pcorpet the usecase you mention is valid. However, following the pattern used by `ReadConcern` does not preclude you from using this functionality as you intend. When comparing two objects of disparate types, when the `__eq__` method of one type returns `NotImplemented`, the corresponding `__eq__` method of the other object gets invoked automatically. This means you can implement whatever logic you'd like to use to compare your custom class with `WriteConcern` in `YourClass.__eq__` and things will work just fine! Here's an example of this in action:

```
class A(object):
"""This is analogous to your homegrown WriteConcern with some custom __eq__ logic. """
    def __eq__(self, other):
        return True
    
class B(object):
"""This is analogous to pymongo's WriteConcern with the suggested changes. """
    def __eq__(self, other):
        return NotImplemented
    
>>> A() == B()
>>> True
```

prashantmital commented on test/test_write_concern.py:
```diff
@@ -0,0 +1,32 @@
+
+    def test_equality_to_none(self):
+        concern = WriteConcern()
+        none = None
```
Aliasing `None` doesn't really make sense as it is a shared global constant. I would just use `None` directly in the assertions.

pcorpet replied:
It was to silence the pep8 warning (comparing with literal `None` should be done by using `is None`, however it's OK when it's a dynamic var that can have the `None` vaue). I've used a #noqa annotation instead.

pcorpet commented:
Thanks for pushing, you're totally right. I've kept a test to make sure it's indeed possible to have equality in that case.

prashantmital commented on test/test_write_concern.py:
```diff
@@ -0,0 +1,50 @@
+
+    def test_equality_to_none(self):
+        concern = WriteConcern()
+        self.assertTrue(concern != None)  # noqa
```
Is there a reason we are not using `self.assertEqual` and `self.assertNotEqual` here? It's not a big deal but personally, I am not a fan of using `self.assertTrue` because it doesn't make the test-writers intent obvious (one must inspect the expression to infer what is being tested).

pcorpet replied:
I wanted to explicitly call `!=`, but I've switched to assertNotEqual for the other one (which uses == internally). I've added a comment to make it more obvious.

prashantmital commented on test/test_write_concern.py:
```diff
@@ -0,0 +1,50 @@
+import collections
```
We need to preface this code with the copyright/license note. You can use the following text:

```
# Copyright 2018-present MongoDB, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
```

pcorpet replied:
Done.

I suggest that you add a presubmit check (or a test on Travis) making sure that all files are prefaced with a copyright note, that will save future contributors some time and save you some peace of mind.

prashantmital replied:
I'll look into creating a presubmit check, thanks for your suggestion.

prashantmital commented:
Thanks @pcorpet !

prashantmital commented:
Closes https://jira.mongodb.org/browse/PYTHON-1631.
